streams/src/main/java/org/apache/kafka/streams/kstream/internals/CogroupedKStreamImpl.java

```diff
@@ -0,0 +1,91 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.streams.kstream.internals;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import org.apache.kafka.common.utils.Bytes;
+import org.apache.kafka.streams.kstream.Aggregator;
+import org.apache.kafka.streams.kstream.CogroupedKStream;
+import org.apache.kafka.streams.kstream.Initializer;
+import org.apache.kafka.streams.kstream.KGroupedStream;
+import org.apache.kafka.streams.kstream.KTable;
+import org.apache.kafka.streams.kstream.Materialized;
+import org.apache.kafka.streams.kstream.internals.graph.StreamsGraphNode;
+import org.apache.kafka.streams.state.KeyValueStore;
+
+public class CogroupedKStreamImpl<K, Vout> extends AbstractStream<K, Vout> implements CogroupedKStream<K, Vout> {
+
+    private static final String AGGREGATE_NAME = "COGROUPKSTREAM-AGGREGATE-";
+
+    final private Map<KGroupedStreamImpl<K, ?>, Aggregator<? super K, ? super Object, Vout>> groupPatterns;
+    final private CogroupedStreamAggregateBuilder<K, Vout> aggregateBuilder;
+
+    CogroupedKStreamImpl(final String name,
+                         final Set<String> sourceNodes,
+                         final StreamsGraphNode streamsGraphNode,
+                         final InternalStreamsBuilder builder) {
+        super(name, null, null, sourceNodes, streamsGraphNode, builder);
```

Comment:
Because it is not obvious from which stream to get the serde so either  override from the materialized object, or use the default from the config.

Also we do not pass the value Serdes because we cannot do so safely because each streams can have a different value type

Reply:
Ah. I missed that we have only only `CogroupedKStreamImpl` object (my mental model was that we have one for each input stream).

What's unclear to me atm (maybe I need to do more detailed review) is, how repartitioning works? For that case, when do we insert a "source" node that is reading from the repartition topic, and where does the source node get the `Serde` information from? We could also have multiple independent repartition steps for different input streams.

Reply:
we had a couple long conversations about this. We resolved that the Serdes would come from the Materialized when aggregated or the default for that type

Reply:
I think there is some misunderstanding. I understand that we cannot pass in `Serdes` to the superclass, but we still need to track them. Assume, the following programs:
```
KStream<String, Integer> stream1 = builder.stream(...).map(...);
KStream<String, Long> stream2 = builder.stream(...).map(...);

KTable<String, Float> result = stream1.groupByKey(Grouped.with(...))
    .cogroup(stream2.groupByKey(Grouped.with(...)))
    .aggregate(...);
```

Because we do a `map()` the serdes from `builder.stream()` cannot be forwarded and we also need to repartition each stream before we can compute the `aggregation()`. Hence, the use needs to pass in the Serdes for the new types (ie, result types of `map()` via `Grouped()` into each operation -- note, that both value types can be different. Because we apply repartitioning lazy, we would only insert it when we call `aggregate()`. Hence, the topology we need to generate later must be:
```
source1 -> map1 -> sink1 -> repartitionTopic1 -> source3 -> aggProcessor1 --+
                                                                            |
                                                                            + --> mergeNode
                                                                            |
source2 -> map2 -> sink2 -> repartitionTopic2 -> source4 -> aggProcessor2 --+
```

For both repartition steps, we need to use different Serdes (as provided by each `Grouped` parameter) -- hence, the `GroupedKStream` must forward the corresponding Serdes to `CogroupedKStreamImpl` because this class will add all corresponding `StreamGraphNodes` (including the ones to repartition the data if necessary. Compare https://github.com/apache/kafka/blob/trunk/streams/src/main/java/org/apache/kafka/streams/kstream/internals/GroupedStreamAggregateBuilder.java#L84-L99

Not sure atm what the best way is to implement this. Maybe just adding
```
final private Map<KGroupedStreamImpl<K, ?>, Serde> keySerde;
final private Map<KGroupedStreamImpl<K, ?>, Serde> valueSerde;
```

to `CogroupedKStreamImpl` (ie, next to `groupPatterns`) might be sufficient. Maybe there is a more elegant way? Maybe we can introduce a small helper Pojo to wrap `Aggregator` + both `Serdes` and have only one `Map` instead (to avoid scattering the state we track into different data structures)

The Serdes from `Materialized` only apply to the _output_ value type that can be different to the input value types (in fact, input value types can be different to each other, too).